starfish/codebook.py

```diff
@@ -371,52 +460,27 @@ def decode_euclidean(self, intensities: IntensityTable) -> IntensityTable:
             qualities
 
         """
+        # normalize both the intensities and the codebook
+        norm_intensities, norms = self._normalize_features(intensities, norm=norm)
+        norm_codes, _ = self._normalize_features(self, norm=norm)
 
-        def _min_euclidean_distance(observation: xr.DataArray, codes: Codebook) -> np.ndarray:
-            """find the code with the closest euclidean distance to observation
-
-            Parameters
-            ----------
-            observation : xr.DataArray
-                2-dimensional DataArray of shape (n_ch, n_hyb)
-            codes :
-                Codebook containing codes to compare to observation
-
-            Returns
-            -------
-            np.ndarray :
-                1-d vector containing the distance of each code to observation
+        # mask low intensity features
+        intensity_mask = np.where(norms < min_intensity)[0]
 
-            """
-            squared_diff = (codes - observation) ** 2
-            code_distances = np.sqrt(squared_diff.sum((Indices.CH, Indices.HYB)))
-            # order of codes changes here (automated sorting on the reshaping?)
-            return code_distances
+        qualities, gene_ids = self._approximate_nearest_code(norm_codes, norm_intensities)
 
-        # normalize both the intensities and the codebook
-        norm_intensities = intensities.groupby(IntensityTable.Constants.FEATURES.value).apply(
-            lambda x: x / x.sum())
-        norm_codes = self.groupby(Codebook.Constants.GENE.value).apply(lambda x: x / x.sum())
-
-        # calculate pairwise euclidean distance between codes and features
-        func = functools.partial(_min_euclidean_distance, codes=norm_codes)
-        distances = norm_intensities.groupby(IntensityTable.Constants.FEATURES.value).apply(func)
-
-        # calculate quality of each decoded spot
-        qualities = 1 - distances.min(Codebook.Constants.GENE.value)
-        qualities_index = pd.Index(qualities)
-
-        # identify genes associated with closest codes
-        closest_code_index = distances.argmin(Codebook.Constants.GENE.value)
-        gene_ids = distances.indexes[
-            Codebook.Constants.GENE.value].values[closest_code_index.values]
+        exceeds_distance = np.where(qualities < 1 - max_distance)[0]
```

Comment:
careful about signs and inequality directions after you make he switch from quality to distance.